### PR TITLE
Undertone: Bidder implementation

### PR DESCRIPTION
**Pull request by tamarm-perion**

### CODE REVIEW

File: sample/prebid-config.yaml [hunk 1]
```diff
@@ -4,6 +4,8 @@ adapters:
     enabled: true
     XAPI.Username: user1
     XAPI.Password: password1
+  undertone:
```
Comment (SerhiiNahornyi) [inline sample/prebid-config.yaml:7]: We don't add newly created bidder config here. Pls remove

File: src/main/java/org/prebid/server/bidder/undertone/UndertoneBidder.java [hunk 2]
```diff
@@ -0,0 +1,217 @@
...
+
+        final ExtRequestPrebid extRequestPrebid = ExtRequestPrebid.builder()
+                .bidderparams(BIDDER_PARAMS)
+                .build();
+
+        final BidRequest.BidRequestBuilder bidRequestBuilder = bidRequest.toBuilder()
+                .imp(imps)
+                .ext(ExtRequest.of(extRequestPrebid));
+
+        if (bidRequest.getSite() != null) {
```
Comment (SerhiiNahornyi) [inline src/main/java/org/prebid/server/bidder/undertone/UndertoneBidder.java:105]: Pls extract `bidRequest.getSite()` to separate variable, as you use it more then once

File: src/main/java/org/prebid/server/bidder/undertone/UndertoneBidder.java [hunk 3]
```diff
@@ -0,0 +1,217 @@
...
+            bidRequestBuilder.site(bidRequest.getSite()
+                    .toBuilder()
+                    .publisher(publisher)
+                    .build());
+        } else {
```
Comment (SerhiiNahornyi) [inline src/main/java/org/prebid/server/bidder/undertone/UndertoneBidder.java:110]: Pls extract `bidRequest.getApp()` to separate variable, as you use it more then once

File: src/main/java/org/prebid/server/bidder/undertone/UndertoneBidder.java [hunk 4]
```diff
@@ -0,0 +1,217 @@
...
+    }
+
+    private static ObjectNode makeBidderParams() {
+        final ObjectNode objectNode = new ObjectMapper().createObjectNode();
+        objectNode.put("id", ADAPTER_ID);
+        objectNode.put("version", VERSION);
+        return objectNode;
+    }
+
+    private Publisher makePublisher(BidRequest bidRequest, int publisherId) {
```
Comment (SerhiiNahornyi) [inline src/main/java/org/prebid/server/bidder/undertone/UndertoneBidder.java:127]: You don't need to pass whole bidRequest, just site and app enough

File: src/main/java/org/prebid/server/bidder/undertone/UndertoneBidder.java [hunk 5]
```diff
@@ -0,0 +1,217 @@
...
+    }
+
+    private boolean isValidImp(Imp imp, ExtImpUndertone extImpUndertone) {
+        return imp != null && ObjectUtils.anyNotNull(imp.getVideo(), imp.getBanner())
+                && extImpUndertone != null && extImpUndertone.getPlacementId() != null;
+    }
+
+    private List<BidderBid> extractBids(BidRequest bidRequest, BidResponse bidResponse) {
+        if (bidResponse == null || CollectionUtils.isEmpty(bidResponse.getSeatbid())) {
+            return List.of();
```
Comment (SerhiiNahornyi) [inline src/main/java/org/prebid/server/bidder/undertone/UndertoneBidder.java:175]: Codestyle. For empty collections we use `Collections.emptyList()`

File: src/main/java/org/prebid/server/bidder/undertone/UndertoneBidder.java [hunk 6]
```diff
@@ -0,0 +1,217 @@
...
+                .filter(Objects::nonNull)
+                .flatMap(Collection::stream)
+                .map(bid -> BidderBid.of(bid, getBidType(bid, idImpMap), bidResponse.getCur()))
+                .toList();
+    }
+
+    private Map<String, Imp> getIdImpMap(BidRequest bidRequest) {
+        return bidRequest.getImp()
+                .stream()
+                .collect(Collectors.groupingBy(Imp::getId))
```
Comment (SerhiiNahornyi) [inline src/main/java/org/prebid/server/bidder/undertone/UndertoneBidder.java:196]: Core will not let you have two imps with identical id.
This code is redundant

File: src/main/java/org/prebid/server/bidder/undertone/UndertoneBidder.java [hunk 7]
```diff
@@ -0,0 +1,217 @@
...
+                .entrySet()
+                .stream()
+                .collect(Collectors.toUnmodifiableMap(Map.Entry::getKey, imps -> imps.getValue().get(0)));
```
Comment (SerhiiNahornyi) [inline src/main/java/org/prebid/server/bidder/undertone/UndertoneBidder.java:199]: Collectors.toMap is enough

File: src/main/java/org/prebid/server/proto/openrtb/ext/request/undertone/ExtImpUndertone.java [hunk 8]
```diff
@@ -0,0 +1,17 @@
+package org.prebid.server.proto.openrtb.ext.request.undertone;
+
+import com.fasterxml.jackson.annotation.JsonProperty;
+import lombok.AllArgsConstructor;
+import lombok.Value;
+
+@AllArgsConstructor(staticName = "of")
+@Value
```
Comment (SerhiiNahornyi) [inline src/main/java/org/prebid/server/proto/openrtb/ext/request/undertone/ExtImpUndertone.java:8]: @Value(staticConstructor = "of")

File: src/test/resources/org/prebid/server/it/openrtb2/undertone/test-cache-undertone-request.json [hunk 9]
```diff
@@ -0,0 +1,18 @@
+{
```
Comment (SerhiiNahornyi) [inline src/test/resources/org/prebid/server/it/openrtb2/undertone/test-cache-undertone-request.json:1]: Redundant file

File: src/test/resources/org/prebid/server/it/openrtb2/undertone/test-cache-undertone-response.json [hunk 10]
```diff
@@ -0,0 +1,7 @@
+{
```
Comment (SerhiiNahornyi) [inline src/test/resources/org/prebid/server/it/openrtb2/undertone/test-cache-undertone-response.json:1]: Redundant file

File: src/test/resources/org/prebid/server/it/test-application.properties [hunk 11]
```diff
@@ -322,6 +322,8 @@ adapters.zeroclickfraud.enabled=true
 adapters.zeroclickfraud.endpoint=http://{{Host}}/zeroclickfraud-exchange?sid={{SourceId}}
 adapters.aax.enabled=true
 adapters.aax.endpoint=http://localhost:8090/aax-exchange
+adapters.undertone.enabled=true
```
Comment (SerhiiNahornyi) [inline src/test/resources/org/prebid/server/it/test-application.properties:325]: We prefer to have properties in alphabetical order, pls set it in appropriate place

**Review by SerhiiNahornyi — CHANGES_REQUESTED** [review]

File: src/main/java/org/prebid/server/bidder/undertone/UndertoneBidder.java (hunk 3 above)
Comment (tamarm-perion) [reply src/main/java/org/prebid/server/bidder/undertone/UndertoneBidder.java:114]: `bidRequest.getApp()` is used once inside the else statement

File: src/main/java/org/prebid/server/bidder/undertone/UndertoneBidder.java (hunk 6 above)
Comment (tamarm-perion) [reply src/main/java/org/prebid/server/bidder/undertone/UndertoneBidder.java:200]: `getBidType` method uses this map, would you like us to change the implementation?

**Review by SerhiiNahornyi — APPROVED** [review]

**Pull request merged**